include/SH3/system/glvertarray.hpp

```diff
@@ -0,0 +1,287 @@
+#include "GL/gl.h"
+
+#include "SH3/system/assert.hpp"
+#include "SH3/system/glbuffer.hpp"
+
+namespace sh3_gl{
+
+    /**
+     *  Contains actual functionality of our VAO (@ref Bind, @ref Create etc).
+     *
+     *  @note Should <i>not</i> be directly instantiated, use @ref vao instead.
+     */
+    struct finalvao
+    {
+    protected:
+
+        /**
+         *  Non indexed constructor.
+         *
+         *  @param verts Reference to our @ref buffer_object filled with vertex data.
+         */
+        finalvao(const buffer_object& verts):ibo(nullptr), vertices(verts)
+        {
+            Create();
+        };
+
+        /**
+         *  Indexed constructor.
+         *
+         *  @param ibo      Reference to our @ref buffer_object filled with vertex <i>indices</i> instead of vertex data.
+         *  @param verts    Reference to our @ref buffer_object filled with vertex data.
+         */
+        finalvao(const buffer_object &ibo, const buffer_object& verts):ibo(&ibo), vertices(verts)
+        {
+            Create();
+        };
+
+        //No virtual dtor, because finalvao should not be stored directly, i.e. this should only ever be used as a reference.
+
+    private:
+        //Prevent dynamic instantiation since we have no virtual dtor.
+        static void* operator new(std::size_t) = delete;
+        static void* operator new[](std::size_t) = delete;
+        static void operator delete(void*) = delete;
+        static void operator delete[](void*) = delete;
+
+    public:
+        /**
+         *  Create a Vertex Attribute Object
+         */
+        void Create();
+
+        /**
+         *  Realease our VAO.
+         */
+        void Release();
+
+        /**
+         *  Bind this VAO.
+         */
+        void Bind();
+
+        /**
+         * Unind this VAO.
+         */
+        void Unbind();
+
+        /**
+         * Get the ID of this VAO.
+         *
+         *  @return @ref id of this VAO.
+         */
+        GLuint GetID() const {return id;}
+
+        /**
+         *  Draw our VAO.
+         */
+        void Draw();
+
+    protected:
+        GLuint  id;                         /**< Our VAO's identification number. **/
+        const buffer_object& vertices;      /**< Vertex Buffer Object */
+
+        //FIXME: std::optional<buffer_object&> would be nicer
+        const buffer_object *const ibo;     /**< (Optional) Buffer for indexed vertices. **/
+    };
+
+    /**
+     *  A VAO without indexed vertices.
+     *
+     *  You probably want to use @ref vao.
+     *
+     *  @tparam numbuffers Number of buffers that should be reserved in @ref buffers.
+     */
+    template<std::size_t numbuffers>
+    struct mutablevao : public finalvao
+    {
+    protected:
+
+        /**
+         *  Our VAO that inherits from @ref finalvao. Contains a list of all our @ref buffer_object structures.
+         *
+         *  @param vertices Reference to our vertices @ref buffer_object. This gets passed
```

Comment:
" This gets passed to @ref": this sentence is missing an end. Also, unnecessary implementation detail. Just omit it.